File: src/main/java/com/gmail/borlandlp/minigamesdtools/gui/inventory/InventoryListener.java
package com.gmail.borlandlp.minigamesdtools.gui.inventory;

import com.gmail.borlandlp.minigamesdtools.Debug;
import com.gmail.borlandlp.minigamesdtools.MinigamesDTools;
import com.gmail.borlandlp.minigamesdtools.creator.AbstractDataProvider;
import com.gmail.borlandlp.minigamesdtools.creator.DataProvider;
import com.gmail.borlandlp.minigamesdtools.events.PlayerRequestOpenInvGUIEvent;
import com.gmail.borlandlp.minigamesdtools.gui.inventory.slots.InventorySlot;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryCloseEvent;
import org.bukkit.event.player.PlayerQuitEvent;

import java.time.Instant;
import java.util.HashMap;

public class InventoryListener implements Listener {
    private HashMap<String, Long> playersClickLog = new HashMap<>();

    @EventHandler
    public void onInvRequest(PlayerRequestOpenInvGUIEvent event) {
        if(!MinigamesDTools.getInstance().getInventoryGUICreatorHub().containsRouteId2Creator(event.getPageID())) {
            Debug.print(Debug.LEVEL.WARNING, "Player[name:" + event.getPlayer().getName() + "] requested unknown Page[ID:" + event.getPageID() + "]");
            return;
        }

        DrawableInventory drawableInventory = null;
        try {
            AbstractDataProvider dataProvider = new DataProvider();
            dataProvider.set("player", event.getPlayer());
            drawableInventory = MinigamesDTools.getInstance().getInventoryGUICreatorHub().createInventory(event.getPageID(), dataProvider);
        } catch (Exception e) {
            e.printStackTrace();
        }

        if(drawableInventory != null) {
            MinigamesDTools.getInstance().getInventoryGUI_API().show(drawableInventory, event.getPlayer());
        } else {
            Debug.print(Debug.LEVEL.NOTICE, "Internal error. DrawableInventory cant be null.");
        }
    }

    @EventHandler(
            priority = EventPriority.LOW,
            ignoreCancelled = true
    )
    public void onCloseInv(InventoryCloseEvent event) {
        if(MinigamesDTools.getInstance().getInventoryGUI_API().isView(((Player)event.getPlayer()))) {
            MinigamesDTools.getInstance().getInventoryGUI_API().unregister((Player)event.getPlayer());
        }
    }

    @EventHandler(
            priority = EventPriority.LOW
    )
    public void onPlayerQuitEvent(PlayerQuitEvent event) {
        if(MinigamesDTools.getInstance().getInventoryGUI_API().isView((event.getPlayer()))) {
            MinigamesDTools.getInstance().getInventoryGUI_API().close(event.getPlayer());
        }
    }

    @EventHandler(priority = EventPriority.LOW)
    public void onPlayerKickEvent(PlayerQuitEvent event) {
        if(MinigamesDTools.getInstance().getInventoryGUI_API().isView((event.getPlayer()))) {
            MinigamesDTools.getInstance().getInventoryGUI_API().close(event.getPlayer());
        }
    }

    @EventHandler(
            priority = EventPriority.LOW,
            ignoreCancelled = true
    )
    public void onInvClickItem(InventoryClickEvent event) {
        if(event.getClickedInventory() == null) {
            return;
        } else if(MinigamesDTools.getInstance().getInventoryGUI_API().isView((Player) event.getWhoClicked())) {

            if(event.getWhoClicked().getOpenInventory().getTopInventory() != event.getClickedInventory()) {
                return;
            }

            event.setCancelled(true);

            if(event.getCurrentItem().getType().name().equalsIgnoreCase("air")) {
                return;
            }

            DrawableInventory drawableInventory = MinigamesDTools.getInstance().getInventoryGUI_API().getInventory((Player)event.getWhoClicked());

            String playerName = event.getWhoClicked().getName();
            if(!this.playersClickLog.containsKey(playerName) || (this.playersClickLog.get(playerName) + 1) <= Instant.now().getEpochSecond()) {
                this.playersClickLog.put(playerName, Instant.now().getEpochSecond());
            } else {
                event.getWhoClicked().sendMessage("Пожалуйста, не кликайте так часто! :( Минимальное время между кликами - 1 секунда.");
                this.playersClickLog.put(playerName, Instant.now().getEpochSecond());
                return;
            }

            InventorySlot slot = drawableInventory.getSlot(event.getSlot());
            if(slot != null) {
                slot.performClick((Player) event.getWhoClicked());
            }

        } else if((event.getWhoClicked()).getOpenInventory() != null && MinigamesDTools.getInstance().getInventoryGUI_API().isView((Player) event.getWhoClicked()) && event.isShiftClick()) {
            event.setCancelled(true);
        }
    }
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/gui/inventory/InventoryListener.kt
package com.gmail.borlandlp.minigamesdtools.gui.inventory

import com.gmail.borlandlp.minigamesdtools.Debug
import com.gmail.borlandlp.minigamesdtools.MinigamesDTools
import com.gmail.borlandlp.minigamesdtools.creator.AbstractDataProvider
import com.gmail.borlandlp.minigamesdtools.creator.DataProvider
import com.gmail.borlandlp.minigamesdtools.events.PlayerRequestOpenInvGUIEvent
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.inventory.InventoryCloseEvent
import org.bukkit.event.player.PlayerQuitEvent
import java.time.Instant
import java.util.*

class InventoryListener : Listener {
    private val playersClickLog = HashMap<String, Long?>()
    @EventHandler
    fun onInvRequest(event: PlayerRequestOpenInvGUIEvent) {
        if (!MinigamesDTools.instance!!.inventoryGUICreatorHub!!.containsRouteId2Creator(event.pageID)) {
            Debug.print(
                Debug.LEVEL.WARNING,
                "Player[name:" + event.player.name + "] requested unknown Page[ID:" + event.pageID + "]"
            )
            return
        }
        var drawableInventory: DrawableInventory? = null
        try {
            val dataProvider: AbstractDataProvider = DataProvider()
            dataProvider["player"] = event.player
            drawableInventory = MinigamesDTools.instance!!.inventoryGUICreatorHub!!
                .createInventory(event.pageID, dataProvider)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        if (drawableInventory != null) {
            MinigamesDTools.instance!!.inventoryGUI_API!!.show(drawableInventory, event.player)
        } else {
            Debug.print(
                Debug.LEVEL.NOTICE,
                "Internal error. DrawableInventory cant be null."
            )
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    fun onCloseInv(event: InventoryCloseEvent) {
        if (MinigamesDTools.instance!!.inventoryGUI_API!!.isView((event.player as Player))) {
            MinigamesDTools.instance!!.inventoryGUI_API!!.unregister((event.player as Player))
        }
    }

    @EventHandler(priority = EventPriority.LOW)
    fun onPlayerQuitEvent(event: PlayerQuitEvent) {
        if (MinigamesDTools.instance!!.inventoryGUI_API!!.isView(event.player)) {
            MinigamesDTools.instance!!.inventoryGUI_API!!.close(event.player)
        }
    }

    @EventHandler(priority = EventPriority.LOW)
    fun onPlayerKickEvent(event: PlayerQuitEvent) {
        if (MinigamesDTools.instance!!.inventoryGUI_API!!.isView(event.player)) {
            MinigamesDTools.instance!!.inventoryGUI_API!!.close(event.player)
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    fun onInvClickItem(event: InventoryClickEvent) {
        if (event.clickedInventory == null) {
            return
        } else if (MinigamesDTools.instance!!.inventoryGUI_API!!.isView((event.whoClicked as Player))) {
            if (event.whoClicked.openInventory.topInventory !== event.clickedInventory) {
                return
            }
            event.isCancelled = true
            if (event.currentItem.type.name.equals("air", ignoreCase = true)) {
                return
            }
            val drawableInventory =
                MinigamesDTools.instance!!.inventoryGUI_API!!.getInventory((event.whoClicked as Player))
            val playerName = event.whoClicked.name
            if (!playersClickLog.containsKey(playerName) || playersClickLog[playerName]!! + 1 <= Instant.now().epochSecond) {
                playersClickLog[playerName] = Instant.now().epochSecond
            } else {
                event.whoClicked
                    .sendMessage("Пожалуйста, не кликайте так часто! :( Минимальное время между кликами - 1 секунда.")
                playersClickLog[playerName] = Instant.now().epochSecond
                return
            }
            val slot = drawableInventory.getSlot(event.slot)
            slot.performClick(event.whoClicked as Player)
        } else if (event.whoClicked.openInventory != null && MinigamesDTools.instance!!.inventoryGUI_API!!.isView(
                (event.whoClicked as Player)
            ) && event.isShiftClick
        ) {
            event.isCancelled = true
        }
    }
}